plugins/bigquery/dbt/adapters/bigquery/impl.py

```diff
@@ -26,6 +29,93 @@
 
 import time
 import agate
+import re
+
+
+BQ_INTEGER_RANGE_NOT_SUPPORTED = f"""
+BigQuery integer range partitioning is only supported by the
+`partition_by` config, which accepts a dictionary.
+
+See: {dbt.links.BigQueryNewPartitionBy}
+"""
+
+
+@dataclass
+class PartitionConfig(Dict[str, Any], JsonSchemaMixin):
+    field: str
+    data_type: str
+    range: Optional[Dict[str, any]] = None
+
+    def render(self, alias: Optional[str] = None):
+        column: str = self.field
+        if alias:
+            column = f'{alias}.{self.field}'
+
+        if self.data_type in ('timestamp', 'datetime'):
+            return f'date({column})'
+        else:
+            return column
+
+    @classmethod
+    def _parse(cls, raw_partition_by) -> Optional['PartitionConfig']:
+        if isinstance(raw_partition_by, dict):
+            if raw_partition_by.get('field'):
+                if raw_partition_by.get('data_type'):
+                    return cls(**raw_partition_by)
+                else:  # assume date type as default
+                    return cls(**raw_partition_by, data_type='date')
+            else:
+                dbt.exceptions.raise_compiler_error(
+                    'Config `partition_by` is missing required item `field`'
+                )
+
+        elif isinstance(raw_partition_by, str):
+            raw_partition_by = raw_partition_by.strip()
+            if 'range_bucket' in raw_partition_by.lower():
+                dbt.exceptions.raise_compiler_error(
+                    BQ_INTEGER_RANGE_NOT_SUPPORTED
+                )
+
+            elif raw_partition_by.lower().startswith('date('):
+                matches = re.match(r'date\((.+)\)', raw_partition_by,
+                                   re.IGNORECASE)
+                if not matches:
+                    dbt.exceptions.raise_compiler_error(
+                        f"Specified partition_by '{raw_partition_by}' "
+                        "is not parseable")
+
+                partition_by = matches.group(1)
+                data_type = 'timestamp'
+
+            else:
+                partition_by = raw_partition_by
+                data_type = 'date'
+
+            inferred_partition_by = {
+                'field': partition_by,
+                'data_type': data_type
+            }
```

Comment:
How about:

```python
inferred_partition_by = cls(field=partition_by, data_type=data_type)
dbt.deprecations.warn(
    'bq-partition-by-string',
    raw_partition_by=raw_partition_by,
    inferred_partition_by=inferred_partition_by
)
return inferred_partition_by
```

Setting up a kwargs dictionary just to expand it seems unnecessary!